File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/RegexTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.beam.sdk.transforms;

import java.io.Serializable;
import java.util.Arrays;
import java.util.List;
import org.apache.beam.sdk.testing.NeedsRunner;
import org.apache.beam.sdk.testing.PAssert;
import org.apache.beam.sdk.testing.TestPipeline;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.PCollection;
import org.junit.Rule;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Tests for {@link Regex}. */
@RunWith(JUnit4.class)
public class RegexTest implements Serializable {

  @Rule public final transient TestPipeline p = TestPipeline.create();

  @Test
  @Category(NeedsRunner.class)
  public void testFind() {
    PCollection<String> output =
        p.apply(Create.of("aj", "xj", "yj", "zj")).apply(Regex.find("[xyz]"));

    PAssert.that(output).containsInAnyOrder("x", "y", "z");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testFindGroup() {
    PCollection<String> output =
        p.apply(Create.of("aj", "xj", "yj", "zj")).apply(Regex.find("([xyz])", 1));

    PAssert.that(output).containsInAnyOrder("x", "y", "z");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testFindNone() {
    PCollection<String> output = p.apply(Create.of("a", "b", "c", "d")).apply(Regex.find("[xyz]"));

    PAssert.that(output).empty();
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testFindNameGroup() {
    PCollection<String> output =
        p.apply(Create.of("aj", "xj", "yj", "zj"))
            .apply(Regex.find("(?<namedgroup>[xyz])", "namedgroup"));

    PAssert.that(output).containsInAnyOrder("x", "y", "z");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testFindAllGroups() {
    PCollection<List<String>> output =
        p.apply(Create.of("aj", "xjx", "yjy", "zjz")).apply(Regex.findAll("([xyz])j([xyz])"));

    PAssert.that(output)
        .containsInAnyOrder(
            Arrays.asList("xjx", "x", "x"),
            Arrays.asList("yjy", "y", "y"),
            Arrays.asList("zjz", "z", "z"));
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testFindNameNone() {
    PCollection<String> output =
        p.apply(Create.of("a", "b", "c", "d"))
            .apply(Regex.find("(?<namedgroup>[xyz])", "namedgroup"));

    PAssert.that(output).empty();
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testKVFind() {

    PCollection<KV<String, String>> output =
        p.apply(Create.of("a b c")).apply(Regex.findKV("a (b) (c)", 1, 2));

    PAssert.that(output).containsInAnyOrder(KV.of("b", "c"));
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testKVFindNone() {

    PCollection<KV<String, String>> output =
        p.apply(Create.of("x y z")).apply(Regex.findKV("a (b) (c)", 1, 2));

    PAssert.that(output).empty();
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testKVFindName() {

    PCollection<KV<String, String>> output =
        p.apply(Create.of("a b c"))
            .apply(Regex.findKV("a (?<keyname>b) (?<valuename>c)", "keyname", "valuename"));

    PAssert.that(output).containsInAnyOrder(KV.of("b", "c"));
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testKVFindNameNone() {

    PCollection<KV<String, String>> output =
        p.apply(Create.of("x y z"))
            .apply(Regex.findKV("a (?<keyname>b) (?<valuename>c)", "keyname", "valuename"));

    PAssert.that(output).empty();
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testMatches() {

    PCollection<String> output =
        p.apply(Create.of("a", "x", "y", "z")).apply(Regex.matches("[xyz]"));

    PAssert.that(output).containsInAnyOrder("x", "y", "z");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testMatchesNone() {

    PCollection<String> output =
        p.apply(Create.of("a", "b", "c", "d")).apply(Regex.matches("[xyz]"));

    PAssert.that(output).empty();
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testMatchesGroup() {

    PCollection<String> output =
        p.apply(Create.of("a", "x xxx", "x yyy", "x zzz")).apply(Regex.matches("x ([xyz]*)", 1));

    PAssert.that(output).containsInAnyOrder("xxx", "yyy", "zzz");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testMatchesName() {

    PCollection<String> output =
        p.apply(Create.of("a", "x xxx", "x yyy", "x zzz"))
            .apply(Regex.matches("x (?<namedgroup>[xyz]*)", "namedgroup"));

    PAssert.that(output).containsInAnyOrder("xxx", "yyy", "zzz");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testMatchesNameNone() {

    PCollection<String> output =
        p.apply(Create.of("a", "b", "c", "d"))
            .apply(Regex.matches("x (?<namedgroup>[xyz]*)", "namedgroup"));

    PAssert.that(output).empty();
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testAllMatches() {

    PCollection<List<String>> output =
        p.apply(Create.of("a x", "x x", "y y", "z z")).apply(Regex.allMatches("([xyz]) ([xyz])"));

    PAssert.that(output)
        .containsInAnyOrder(
            Arrays.asList("x x", "x", "x"),
            Arrays.asList("y y", "y", "y"),
            Arrays.asList("z z", "z", "z"));
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testKVMatches() {

    PCollection<KV<String, String>> output =
        p.apply(Create.of("a b c")).apply(Regex.matchesKV("a (b) (c)", 1, 2));

    PAssert.that(output).containsInAnyOrder(KV.of("b", "c"));
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testKVMatchesNone() {

    PCollection<KV<String, String>> output =
        p.apply(Create.of("x y z")).apply(Regex.matchesKV("a (b) (c)", 1, 2));
    PAssert.that(output).empty();
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testKVMatchesName() {

    PCollection<KV<String, String>> output =
        p.apply(Create.of("a b c"))
            .apply(Regex.findKV("a (?<keyname>b) (?<valuename>c)", "keyname", "valuename"));

    PAssert.that(output).containsInAnyOrder(KV.of("b", "c"));
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testKVMatchesNameNone() {

    PCollection<KV<String, String>> output =
        p.apply(Create.of("x y z"))
            .apply(Regex.findKV("a (?<keyname>b) (?<valuename>c)", "keyname", "valuename"));
    PAssert.that(output).empty();
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testReplaceAll() {

    PCollection<String> output =
        p.apply(Create.of("xj", "yj", "zj")).apply(Regex.replaceAll("[xyz]", "new"));

    PAssert.that(output).containsInAnyOrder("newj", "newj", "newj");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testReplaceAllMixed() {

    PCollection<String> output =
        p.apply(Create.of("abc", "xj", "yj", "zj", "def")).apply(Regex.replaceAll("[xyz]", "new"));

    PAssert.that(output).containsInAnyOrder("abc", "newj", "newj", "newj", "def");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testReplaceFirst() {

    PCollection<String> output =
        p.apply(Create.of("xjx", "yjy", "zjz")).apply(Regex.replaceFirst("[xyz]", "new"));

    PAssert.that(output).containsInAnyOrder("newjx", "newjy", "newjz");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testReplaceFirstMixed() {

    PCollection<String> output =
        p.apply(Create.of("abc", "xjx", "yjy", "zjz", "def"))
            .apply(Regex.replaceFirst("[xyz]", "new"));

    PAssert.that(output).containsInAnyOrder("abc", "newjx", "newjy", "newjz", "def");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testSplits() {

    PCollection<String> output =
        p.apply(Create.of("The  quick   brown fox jumps over    the lazy dog"))
            .apply(Regex.split("\\W+"));

    PAssert.that(output)
        .containsInAnyOrder("The", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testSplitsWithEmpty() {

    PCollection<String> output =
        p.apply(Create.of("The  quick   brown fox jumps over    the lazy dog"))
            .apply(Regex.split("\\s", true));

    PAssert.that(output)
        .containsInAnyOrder(
            "The", "", "quick", "brown", "", "", "fox", "jumps", "over", "", "", "", "the", "lazy",
            "dog");
    p.run();
  }

  @Test
  @Category(NeedsRunner.class)
  public void testSplitsWithoutEmpty() {

    PCollection<String> output =
        p.apply(Create.of("The  quick   brown fox jumps over    the lazy dog"))
            .apply(Regex.split("\\s", false));

    PAssert.that(output)
        .containsInAnyOrder("The", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog");
    p.run();
  }
}

File: sdks/python/apache_beam/examples/snippets/transforms/elementwise/regex_test.py
# coding=utf-8
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# pytype: skip-file
# pylint:disable=line-too-long

import unittest

import mock

from apache_beam.examples.snippets.util import assert_matches_stdout
from apache_beam.testing.test_pipeline import TestPipeline

from . import regex_all_matches
from . import regex_find
from . import regex_find_all
from . import regex_find_kv
from . import regex_matches
from . import regex_matches_kv
from . import regex_replace_all
from . import regex_replace_first
from . import regex_split


def check_matches(actual):
  expected = '''[START plants_matches]
🍓, Strawberry, perennial
🥕, Carrot, biennial
🍆, Eggplant, perennial
🍅, Tomato, annual
🥔, Potato, perennial
[END plants_matches]'''.splitlines()[1:-1]
  assert_matches_stdout(actual, expected)


def check_all_matches(actual):
  expected = '''[START plants_all_matches]
['🍓, Strawberry, perennial', '🍓', 'Strawberry', 'perennial']
['🥕, Carrot, biennial', '🥕', 'Carrot', 'biennial']
['🍆, Eggplant, perennial', '🍆', 'Eggplant', 'perennial']
['🍅, Tomato, annual', '🍅', 'Tomato', 'annual']
['🥔, Potato, perennial', '🥔', 'Potato', 'perennial']
[END plants_all_matches]'''.splitlines()[1:-1]
  assert_matches_stdout(actual, expected)


def check_matches_kv(actual):
  expected = '''[START plants_matches_kv]
('🍓', '🍓, Strawberry, perennial')
('🥕', '🥕, Carrot, biennial')
('🍆', '🍆, Eggplant, perennial')
('🍅', '🍅, Tomato, annual')
('🥔', '🥔, Potato, perennial')
[END plants_matches_kv]'''.splitlines()[1:-1]
  assert_matches_stdout(actual, expected)


def check_find_all(actual):
  expected = '''[START plants_find_all]
['🍓, Strawberry, perennial']
['🥕, Carrot, biennial']
['🍆, Eggplant, perennial', '🍌, Banana, perennial']
['🍅, Tomato, annual', '🍉, Watermelon, annual']
['🥔, Potato, perennial']
[END plants_find_all]'''.splitlines()[1:-1]
  assert_matches_stdout(actual, expected)


def check_find_kv(actual):
  expected = '''[START plants_find_kv]
('🍓', '🍓, Strawberry, perennial')
('🥕', '🥕, Carrot, biennial')
('🍆', '🍆, Eggplant, perennial')
('🍌', '🍌, Banana, perennial')
('🍅', '🍅, Tomato, annual')
('🍉', '🍉, Watermelon, annual')
('🥔', '🥔, Potato, perennial')
[END plants_find_kv]'''.splitlines()[1:-1]
  assert_matches_stdout(actual, expected)


def check_replace_all(actual):
  expected = '''[START plants_replace_all]
🍓,Strawberry,perennial
🥕,Carrot,biennial
🍆,Eggplant,perennial
🍅,Tomato,annual
🥔,Potato,perennial
[END plants_replace_all]'''.splitlines()[1:-1]
  assert_matches_stdout(actual, expected)


def check_replace_first(actual):
  expected = '''[START plants_replace_first]
🍓: Strawberry, perennial
🥕: Carrot, biennial
🍆: Eggplant, perennial
🍅: Tomato, annual
🥔: Potato, perennial
[END plants_replace_first]'''.splitlines()[1:-1]
  assert_matches_stdout(actual, expected)


def check_split(actual):
  expected = '''[START plants_split]
['🍓', 'Strawberry', 'perennial']
['🥕', 'Carrot', 'biennial']
['🍆', 'Eggplant', 'perennial']
['🍅', 'Tomato', 'annual']
['🥔', 'Potato', 'perennial']
[END plants_split]'''.splitlines()[1:-1]
  assert_matches_stdout(actual, expected)


@mock.patch('apache_beam.Pipeline', TestPipeline)
@mock.patch(
    'apache_beam.examples.snippets.transforms.elementwise.regex_matches.print',
    str)
@mock.patch(
    'apache_beam.examples.snippets.transforms.elementwise.regex_all_matches.print',
    str)
@mock.patch(
    'apache_beam.examples.snippets.transforms.elementwise.regex_matches_kv.print',
    str)
@mock.patch(
    'apache_beam.examples.snippets.transforms.elementwise.regex_find.print',
    str)
@mock.patch(
    'apache_beam.examples.snippets.transforms.elementwise.regex_find_all.print',
    str)
@mock.patch(
    'apache_beam.examples.snippets.transforms.elementwise.regex_find_kv.print',
    str)
@mock.patch(
    'apache_beam.examples.snippets.transforms.elementwise.regex_replace_all.print',
    str)
@mock.patch(
    'apache_beam.examples.snippets.transforms.elementwise.regex_replace_first.print',
    str)
@mock.patch(
    'apache_beam.examples.snippets.transforms.elementwise.regex_split.print',
    str)
class RegexTest(unittest.TestCase):
  def test_matches(self):
    regex_matches.regex_matches(check_matches)

  def test_all_matches(self):
    regex_all_matches.regex_all_matches(check_all_matches)

  def test_matches_kv(self):
    regex_matches_kv.regex_matches_kv(check_matches_kv)

  def test_find(self):
    regex_find.regex_find(check_matches)

  def test_find_all(self):
    regex_find_all.regex_find_all(check_find_all)

  def test_find_kv(self):
    regex_find_kv.regex_find_kv(check_find_kv)

  def test_replace_all(self):
    regex_replace_all.regex_replace_all(check_replace_all)

  def test_replace_first(self):
    regex_replace_first.regex_replace_first(check_replace_first)

  def test_split(self):
    regex_split.regex_split(check_split)


if __name__ == '__main__':
  unittest.main()
